engine/src/main/java/org/kigalisim/engine/SingleThreadEngine.java

```diff
@@ -0,0 +1,494 @@
+ * Engine implementation to Java, using BigDecimal for numerical stability.</p>
+ *
+ * @license BSD-3-Clause
+ */
+
+package org.kigalisim.engine;
+
+import java.math.BigDecimal;
+import java.util.HashSet;
+import java.util.Set;
+import org.kigalisim.engine.number.EngineNumber;
+import org.kigalisim.engine.number.UnitConverter;
+import org.kigalisim.engine.state.ConverterStateGetter;
+import org.kigalisim.engine.state.OverridingConverterStateGetter;
+import org.kigalisim.engine.state.Scope;
+import org.kigalisim.engine.state.StreamKeeper;
+import org.kigalisim.engine.state.YearMatcher;
+
+/**
+ * Single-threaded implementation of the Engine interface.
+ *
+ * <p>This implementation provides the core simulation engine functionality
+ * without thread safety considerations. It manages substance streams, equipment
+ * populations, and various calculations related to the Montreal Protocol simulation.</p>
+ */
+public class SingleThreadEngine implements Engine {
+
+  private static final Set<String> STREAM_NAMES = new HashSet<>();
+  private static final boolean OPTIMIZE_RECALCS = true;
+
+  static {
+    STREAM_NAMES.add("priorEquipment");
+    STREAM_NAMES.add("equipment");
+    STREAM_NAMES.add("export");
+    STREAM_NAMES.add("import");
+    STREAM_NAMES.add("manufacture");
+    STREAM_NAMES.add("sales");
+  }
+
+  private final int startYear;
+  private final int endYear;
+  private int currentYear;
+
+  private final ConverterStateGetter stateGetter;
+  private final UnitConverter unitConverter;
+  private final StreamKeeper streamKeeper;
+  private Scope scope;
+
+  /**
+   * Create a new SingleThreadEngine instance.
+   *
+   * @param startYear The starting year of the simulation
+   * @param endYear The ending year of the simulation
+   */
+  public SingleThreadEngine(int startYear, int endYear) {
+    // Ensure start year is less than or equal to end year
+    int startYearRearrange = Math.min(startYear, endYear);
+    int endYearRearrange = Math.max(startYear, endYear);
+
+    this.startYear = startYearRearrange;
+    this.endYear = endYearRearrange;
+    this.currentYear = this.startYear;
+
+    this.stateGetter = new ConverterStateGetter(this);
+    this.unitConverter = new UnitConverter(this.stateGetter);
+    this.streamKeeper = new StreamKeeper(
+        new OverridingConverterStateGetter(this.stateGetter), this.unitConverter);
+    this.scope = new Scope(null, null, null);
+  }
+
+  @Override
+  public int getStartYear() {
+    return startYear;
+  }
+
+  @Override
+  public int getEndYear() {
+    return endYear;
+  }
+
+  @Override
+  public void setStanza(String newStanza) {
+    this.scope = this.scope.getWithStanza(newStanza);
+  }
+
+  @Override
+  public void setApplication(String newApplication) {
+    this.scope = this.scope.getWithApplication(newApplication);
+  }
+
+  @Override
+  public void setSubstance(String newSubstance, Boolean checkValid) {
```

Comment:
Please add a TODO to add an overload instead of having this check for null. It should have the overload with one fewer parameter call the overload with the parameter with a default. Please add similar todos in inline comments for other similar situations in single thread engine. However, we will tackle this in a future PR.